power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,114 +25,240 @@ namespace detail = power_grid_model::optimizer::detail;
 
 using TrafoGraphIdx = Idx;
 using EdgeWeight = int64_t;
-using WeightedTrafo = std::pair<Idx2D, EdgeWeight>;
-using WeightedTrafoList = std::vector<WeightedTrafo>;
-using RankedTransformerGroups = std::vector<std::vector<Idx2D>>;
 constexpr auto infty = std::numeric_limits<Idx>::max();
+constexpr Idx2D unregulated_idx = {-1, -1};
+
+using enum Branch3Side;
+
+constexpr std::array<std::tuple<Branch3Side, Branch3Side>, 3> const branch3_combinations{
+    {{side_1, side_2}, {side_2, side_3}, {side_3, side_1}}};
 
 struct TrafoGraphVertex {
-    bool is_source{}; // is_source = true if the vertex is a source
+    bool is_source{};
 };
 
 struct TrafoGraphEdge {
-    Idx2D pos{};
+    Idx2D regulated_idx{};
     EdgeWeight weight{};
+
+    bool operator==(const TrafoGraphEdge& other) const {
+        return regulated_idx == other.regulated_idx && weight == other.weight;
+    } // thanks boost
+
+    auto operator<=>(const TrafoGraphEdge& other) const {
+        if (auto cmp = weight <=> other.weight; cmp != 0) { // NOLINT(modernize-use-nullptr)
+            return cmp;
+        }
+        if (auto cmp = regulated_idx.group <=> other.regulated_idx.group; cmp != 0) { // NOLINT(modernize-use-nullptr)
```

Comment:
~this is quite a severe clang-tidy error. might be good to investigate why.~ EDIT: bug in clang-tidy. See e.g. https://github.com/llvm/llvm-project/issues/64159#issuecomment-1653873572

probably good in the future to make a comment either inline or on the PR so the reviewer doesn't need to investigate